src/main/cpp/src/timezones.cu

```diff
@@ -0,0 +1,184 @@
+ * @param to_utc whether we are converting to UTC or converting to the timezone
+ * 
+ * @param timestamp input timestamp
+ */
+template <typename timestamp_type>
+struct convert_timestamp_tz_functor {
+  using duration_type = typename timestamp_type::duration;
+
+  lists_column_device_view transitions;
+  size_type tz_index;
+  bool to_utc;
+
+  __device__ timestamp_type operator()(timestamp_type const &timestamp) const {
+    auto const epoch_seconds =
+        static_cast<int64_t>(cuda::std::chrono::duration_cast<cudf::duration_s>(
+          timestamp.time_since_epoch()).count());
+    auto const tz_transitions = cudf::list_device_view{transitions, tz_index};
+    auto const list_size = tz_transitions.size();
+
+    cudf::device_span<int64_t const> transition_times(
+        &(transitions.child()
+              .child(to_utc ? 1 : 0)
+              .data<int64_t>()[tz_transitions.element_offset(0)]),
+        static_cast<size_t>(list_size));
+
+    auto const it = thrust::upper_bound(thrust::seq, transition_times.begin(), transition_times.end(),
+                                  epoch_seconds);
+    auto const idx = static_cast<size_type>(thrust::distance(transition_times.begin(), it));
+
+    auto const list_offset = tz_transitions.element_offset(size_type(idx - 1));
+    auto const utc_offset = cuda::std::chrono::duration_cast<duration_type>(cudf::duration_s{
+        static_cast<int64_t>(transitions.child().child(2).element<int32_t>(list_offset))});
+    return to_utc ? timestamp - utc_offset : timestamp + utc_offset;
+  }
+};
+
+} // namespace
+
+namespace spark_rapids_jni {
+
+std::unique_ptr<column> convert_timestamp_to_utc(cudf::column_view const &input,
+                                                 cudf::table_view const &transitions,
+                                                 size_type tz_index, rmm::cuda_stream_view stream,
+                                                 rmm::mr::device_memory_resource *mr) {
+
+  auto const type = input.type().id();
+  auto const num_rows = input.size();
+
+  // get the fixed transitions
+  auto const ft_cdv_ptr = column_device_view::create(transitions.column(0), stream);
+  lists_column_device_view fixed_transitions = cudf::detail::lists_column_device_view{*ft_cdv_ptr};
+
+  auto results = cudf::make_timestamp_column(input.type(), input.size(),
+                                             cudf::detail::copy_bitmask(input, stream, mr),
+                                             input.null_count(), stream, mr);
+
+  switch (type) {
+    case cudf::type_id::TIMESTAMP_SECONDS:
+      thrust::transform(rmm::exec_policy(stream), 
+                        input.begin<cudf::timestamp_s>(),
+                        input.end<cudf::timestamp_s>(),
+                        results->mutable_view().begin<cudf::timestamp_s>(),
+                        convert_timestamp_tz_functor<cudf::timestamp_s>{fixed_transitions, tz_index, true});
```

Comment:
This snippet is called 6 times. So we can extract it as a template function:
```
template<typename Ts>
auto convert_ts(input, transitions, tz_index, to_utc, stream) {
  auto const fixed_transitions = ....;
  auto result = ....;
  
  thrust::transform(rmm::exec_policy(stream), 
                        input.begin<Ts>(), input.end<Ts>(),
                        results->mutable_view().begin<Ts>(),
                        convert_timestamp_tz_functor<Ts>{fixed_transitions, tz_index, to_utc});
}
```

Then later on, call this 6 times like:
```
switch(type) {
 case TIMESTAMP_SECOND: return convert_ts<cudf::timestamp_s>(input, transitions, tz_index, to_utc, stream);
 case TIMESTAMP_MILLISECOND: return convert_ts<...>();
...
}
```